commit 0ba7536d5d47e4ecf2259a80b207158dc4e711eb
Author: NeilBrown <neilb@cse.unsw.edu.au>
Date:   Mon Nov 7 01:00:26 2005 -0800

    [PATCH] knfsd: Fix some minor sign problems in nfsd/xdr
    
    There are a couple of tests which could possibly be confused by extremely
    large numbers appearing in 'xdr' packets.  I think the closest to an exploit
    you could get would be writing random data from a free page into a file - i.e.
     leak data out of kernel space.
    
    I'm fairly sure they cannot be used for remote compromise.
    
    Signed-off-by: Neil Brown <neilb@suse.de>
    Signed-off-by: Andrew Morton <akpm@osdl.org>
    Signed-off-by: Linus Torvalds <torvalds@osdl.org>

diff --git a/include/linux/sunrpc/svc.h b/include/linux/sunrpc/svc.h
index 5af8800e0ce3..e4086ec8b952 100644
--- a/include/linux/sunrpc/svc.h
+++ b/include/linux/sunrpc/svc.h
@@ -171,7 +171,8 @@ xdr_argsize_check(struct svc_rqst *rqstp, u32 *p)
 {
 	char *cp = (char *)p;
 	struct kvec *vec = &rqstp->rq_arg.head[0];
-	return cp - (char*)vec->iov_base <= vec->iov_len;
+	return cp >= (char*)vec->iov_base
+		&& cp <= (char*)vec->iov_base + vec->iov_len;
 }
 
 static inline int